src/Lykke.Job.CandlesHistoryWriter.Services/HistoryMigration/TradesMigrationManager.cs

```diff
@@ -0,0 +1,139 @@
+                        // To avoid looking through the whole trades batch several times, we derive the
+                        // bigger time intervals from the smaller ones. For example, let's assume that we
+                        // generate 1000 sec candles from our batch of 2381 trades. Then, for calculating
+                        // minute candles we need to iterate only 1000 second candles, and this will give
+                        // us 16 minute candles. Next, we iterate 16 minute candles to get 1 hour candle,
+                        // and so on, instead of looking through 2381 trades for each time interval.
+                        foreach (var interval in Constants.StoredIntervals)
+                        {
+                            // It's important for Constants.StoredIntervals to be ordered by time period increase,
+                            // because we will calculate candles for each period based on previous period candles.
+                            var currentCandles = interval == CandleTimeInterval.Sec
+                                ? new TradesCandleBatch(migrationItem.AssetId, interval, tradesBatch)
+                                : new TradesCandleBatch(migrationItem.AssetId, interval, smallerCandles);
+
+                            ExtendStoredCandles(ref currentCandles);
+
+                            Health.AssetReportItems[migrationItem.AssetId].SummarySavedCandles += currentCandles.CandlesCount;
+
+                            smallerCandles = currentCandles;
+
+                            // Preffered to perform inserting in parralel style.
+                            batchInsertTasks.Add(
+                                _candlesHistoryRepository.InsertOrMergeAsync(
+                                    currentCandles.Candles.Values,
+                                    currentCandles.AssetId,
+                                    TradesCandleBatch.PriceType,
+                                    currentCandles.TimeInterval));
+                        }
+
+                        Health.AssetReportItems[migrationItem.AssetId].SummaryFetchedTrades += batchCount;
+
+                        Task.WaitAll(batchInsertTasks.ToArray());
+
+                        await _log.WriteInfoAsync(nameof(TradesMigrationManager), nameof(Migrate),
+                            $"Batch of {batchCount} records for {migrationItem.AssetId} processed.");
+                    }
+
+                    Health.State = TradesMigrationState.Finished;
+
+                    await _log.WriteInfoAsync(nameof(TradesMigrationManager), nameof(Migrate),
+                        $"Migration for {migrationItem.AssetId} finished. Total records migrated: {Health.AssetReportItems[migrationItem.AssetId].SummaryFetchedTrades}, total candles stored: {Health.AssetReportItems[migrationItem.AssetId].SummarySavedCandles}.");
+                }
+            }
+        }
+
+        private void ExtendStoredCandles(ref TradesCandleBatch current)
```

Comment:
Why current is ref here? Value of current is not changed in this method

Reply:
The "current" itself is not changed in ExtendStoredCandles, but the candles, placed in current, can be modified. The "ref" just signals to anybody that there is some inner modification of the give data in the method, so, he could notice it and investigate the source code for details. But I agree, this is not a standart use of "ref".